File: programmers/level1/완주하지 못한 선수.py
"""
문제 링크: https://programmers.co.kr/learn/courses/30/lessons/42576
문제 접근 방법: 해시
"""

def solution(participant, completion):
    answer = ''

    dic = {}
    for p in participant:
        if dic.get(p):
            dic[p] +=1
        else:
            dic[p] = 1

    for c in completion:
        if dic.get(c):
            dic[c] -=1

    for k,v in dic.items():
        if v != 0:
            answer = k

    return answer

participant = ["leo", "kiki", "eden"]
completion = ["eden", "kiki"]

print(solution(participant,completion))

File: programmers/level1/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        final String[] answer = {""};
        
        Map<String, Integer> dic = new HashMap<>();
        for(String p : participant) {
            if(dic.get(p) == null) {
                dic.put(p,1);
            } else {
                int val = dic.get(p) + 1;
                dic.put(p,val);
            }
        }
        
        for(String c : completion) {
            int val = dic.get(c) - 1;
            dic.put(c, val);
        }
        
        dic.forEach((k,v) -> {
            if(v == 1) answer[0] = k;
        });
        
        
        return answer[0];
    }
}